### PR TITLE
Fix task wizard and implement modify task wizard via graphql

### PR DESCRIPTION
**Pull request by saberlynx**
**What**:

- Implement runModifyTask (Modify Task Wizard) with purely graphql commands.
- Change the date name to current date (in UTC, in ISO string format with seconds) instead of user session timeout.
- Adjust modify task wizard to have similar logic as schedule dialog e.g. only use `startDate` (`start_minute` and `start_hour` will now be inferred from `startDate`, and the time will dynamically adjust based on `startTimezone`)

TODO
- [x] Dynamically generate input mocks
- [x] Change modify task wizard variable name case
- [x] Clean up code

**Why**:

Because modify task wizard can be a frontend only feature.

**How**:

**Checklist**:

- [x] Tests
- [x] [CHANGELOG](https://github.com/greenbone/gsa/blob/master/CHANGELOG.md) Entry

### CODE REVIEW

File: gsa/src/web/graphql/wizards.js [hunk 1]
```diff
@@ -66,8 +68,108 @@ export const useRunQuickFirstScan = () => {
...
+      reschedule,
+      tasks,
+    }) => {
+      start_date.hours(start_hour);
+      start_date.minutes(start_minute);
+
+      const [task] = tasks;
+      const {name, id: taskId, alerts} = task;
+
+      const rightNow = new Date().toISOString();
```
Comment (bjoernricks) [inline gsa/src/web/graphql/wizards.js:98]: I really thing rightNow could be an optional parameter for runModifyTask too.

File: gsa/src/web/graphql/wizards.js [hunk 2]
```diff
@@ -66,8 +68,108 @@ export const useRunQuickFirstScan = () => {
...
+
+      const event = Event.fromData(
+        {startDate: start_date},
+        start_timezone,
```
Comment (bjoernricks) [inline gsa/src/web/graphql/wizards.js:102]:
```suggestion
        {startDate},
        startTimezone,
```

File: gsa/src/web/graphql/wizards.js [hunk 3]
```diff
@@ -66,8 +68,108 @@ export const useRunQuickFirstScan = () => {
...
+        false, // if true, then the event will be scheduled at the given time, but at the user timezone instead. If I create an event at 22:00 Shanghai time, it will always be scheduled at 22:00 CEST (User timezone)
```
Comment (bjoernricks) [inline gsa/src/web/graphql/wizards.js:103]: As mentioned already I am really sure this is wrong. the passed startDate needs to be already in the correct timezone.

File: gsa/src/web/graphql/wizards.js [hunk 4]
```diff
@@ -66,8 +68,108 @@ export const useRunQuickFirstScan = () => {
...
+      );
+      let schedulePromise;
+
+      if (reschedule) {
+        const createScheduleInput = {
+          name: `Schedule for ${name} - ${rightNow}`,
+          icalendar: event.toIcalString(),
+          timezone: start_timezone,
```
Comment (bjoernricks) [inline gsa/src/web/graphql/wizards.js:111]:
```suggestion
          timezone: startTimezone,
```

File: gsa/src/web/graphql/wizards.js [hunk 5]
```diff
@@ -66,8 +68,108 @@ export const useRunQuickFirstScan = () => {
...
+        schedulePromise = Promise.resolve();
+      }
+
+      return schedulePromise.then(resp => {
+        // Should be undefined if no schedule is created
+        const scheduleId = resp?.data?.createSchedule?.id;
+
+        let alertPromise;
+
+        if (alert_email.length > 0) {
```
Comment (bjoernricks) [inline gsa/src/web/graphql/wizards.js:125]: Personally I would check if alert email is also defined to not crash if it isn't.
```suggestion
        if (isDefined(alertEmail) && alertEmail.length > 0) {
```

File: gsa/src/web/graphql/wizards.js [hunk 6]
```diff
@@ -66,8 +68,108 @@ export const useRunQuickFirstScan = () => {
...
+            name: `Email Alert for ${name} - ${rightNow}`,
+            comment: 'Automatically generated by wizard',
+            event: 'TASK_RUN_STATUS_CHANGED',
+            eventData: {
+              status: 'DONE',
+            },
+            method: 'EMAIL',
+            methodData: {
+              to_address: alert_email,
+              from_address: alert_email,
```
Comment (bjoernricks) [inline gsa/src/web/graphql/wizards.js:136]:
```suggestion
              to_address: alertEmail,
              from_address: alertEmail,
```

**Review by bjoernricks — CHANGES_REQUESTED** [review]

File: gsa/src/web/graphql/wizards.js (hunk 6 above)
Comment (saberlynx) [reply gsa/src/web/graphql/wizards.js:121]: I'll refactor modify task wizard to output variable names in camelCase.

File: gsa/src/web/graphql/__tests__/wizards.js [hunk 7]
```diff
@@ -19,38 +19,57 @@
...
+
+beforeAll(() => {
+  global.Date = jest.fn(() => mockDate);
+  global.Date.now = jest.fn(() => mockDate.getTime());
+});
+
+afterAll(() => {
+  global.Date = RealDate;
+  global.Date.now = RealDate.now;
+});
```
Comment (bjoernricks) [inline gsa/src/web/graphql/__tests__/wizards.js:59]: Is this still necessary?

File: gsa/src/web/graphql/__tests__/wizards.js [hunk 8]
```diff
@@ -161,3 +154,199 @@ describe('useRunQuickFirstScan tests', () => {
...
+      tasks: [
+        {
+          name: 'myFirstTask',
+          id: '13579',
+          alerts: [{id: '34567'}],
+        },
+      ],
+    })
+      .then(() => setOk(true))
+      .catch(err => setError(err.message));
```
Comment (bjoernricks) [inline gsa/src/web/graphql/__tests__/wizards.js:178]: you don't need to pass err.message here. The Error class implements to toString() method that returns the message. Therefore the following code would also fit
```suggestion
      .catch(setError);
```

**Review by bjoernricks — CHANGES_REQUESTED** [review]
I would still change `runModifyTask` to allow passing a date. That will be much easier to test then and will not require the mock anymore.

**Comment by saberlynx** [comment]
> I would still change `runModifyTask` to allow passing a date. That will be much easier to test then and will not require the mock anymore.

Ah. So do you mean, allow passing a date so we can pass one when we test, but not necessarily use it in production? (because that would be `rightNow` or something

**Review by bjoernricks — APPROVED** [review]
Really really nice now! Thanks a lot for your patience :-)

**Pull request merged**